service/history/execution/history_builder_test.go

```diff
@@ -1276,6 +1276,253 @@ func (s *historyBuilderSuite) TestHistoryBuilderWorkflowExternalSignaled() {
+	decisionContext := []byte("historybuilder-context")
+	// 4
+	s.addDecisionTaskCompletedEvent(2, 3, decisionContext, identity)
+	decisionResult := []byte("great success")
+	// 5
+	event := s.addWorkflowExecutionCompleted(4, &types.CompleteWorkflowExecutionDecisionAttributes{
+		Result: decisionResult,
+	})
+	s.validateWorkflowExecutionCompleted(event, 5, 4, decisionResult)
+}
+
+func (s *historyBuilderSuite) TestHistoryBuilderTerminateWorkflow() {
+	id := "historybuilder-workflow-id"
+	rid := "historybuilder-test-run-id"
+	wt := "historybuilder-type"
+	tl := "historybuilder-tasklist"
+	identity := "historybuilder-worker"
+	input := []byte("historybuilder-input")
+	execTimeout := int32(60)
+	taskTimeout := int32(10)
+	we := types.WorkflowExecution{
+		WorkflowID: id,
+		RunID:      rid,
+	}
+	partitionConfig := map[string]string{
+		"zone": "dca1",
+	}
+
+	// 1
+	s.addWorkflowExecutionStartedEvent(we, wt, tl, input, execTimeout, taskTimeout, identity, partitionConfig)
+	// 2
+	terminated := s.addWorkflowExecutionTerminated(1, "reason", []byte("a really good reason"), identity)
+	s.validateWorkflowExecutionTerminated(terminated, 2, "reason", []byte("a really good reason"), identity)
+}
+
+func (s *historyBuilderSuite) TestHistoryBuilderCancelWorkflow() {
+	id := "historybuilder-workflow-id"
+	rid := "historybuilder-test-run-id"
+	wt := "historybuilder-type"
+	tl := "historybuilder-tasklist"
+	identity := "historybuilder-worker"
+	input := []byte("historybuilder-input")
+	execTimeout := int32(60)
+	taskTimeout := int32(10)
+	we := types.WorkflowExecution{
+		WorkflowID: id,
+		RunID:      rid,
+	}
+	partitionConfig := map[string]string{
+		"zone": "dca1",
+	}
+
+	// 1
+	s.addWorkflowExecutionStartedEvent(we, wt, tl, input, execTimeout, taskTimeout, identity, partitionConfig)
+	req := &types.HistoryRequestCancelWorkflowExecutionRequest{
+		CancelRequest: &types.RequestCancelWorkflowExecutionRequest{
+			Identity:  "identity",
+			RequestID: "b071cbe8-3a95-4223-a8ac-f308a42db383",
+		},
+	}
+	// 2
+	s.addWorkflowExecutionCancelRequestedEvent("because", req)
+	// 3
+	s.addDecisionTaskScheduledEvent()
+	// 4
+	s.addDecisionTaskStartedEvent(3, tl, identity)
+	// 5
+	s.addDecisionTaskCompletedEvent(3, 4, []byte("context"), identity)
+	// 6
+	event := s.addWorkflowExecutionCanceled(5, &types.CancelWorkflowExecutionDecisionAttributes{
+		Details: []byte("details"),
+	})
+	s.validateWorkflowExecutionCanceled(event, 6, 5, []byte("details"))
+}
+
+func (s *historyBuilderSuite) TestHistoryBuilderTimers() {
+	id := "historybuilder-workflow-id"
+	rid := "historybuilder-test-run-id"
+	wt := "historybuilder-type"
+	tl := "historybuilder-tasklist"
+	identity := "historybuilder-worker"
+	input := []byte("historybuilder-input")
+	execTimeout := int32(60)
+	taskTimeout := int32(10)
+	we := types.WorkflowExecution{
+		WorkflowID: id,
+		RunID:      rid,
+	}
+	partitionConfig := map[string]string{
+		"zone": "dca1",
+	}
+
+	// 1
```

Comment:
Nit: mind expanding on this comment? just a number in comments does not offer much value to readers.

Reply:
This is another oddity of the file that I'm sticking with. These are the event ids, which are allocated to each event in order. Many of the events reference other events, [like on this line](https://github.com/uber/cadence/pull/5977/files/5801149c3bcd91f015a40ec7f2f2a30729d598bd#diff-0fbf138df4faf3cfe428e66630a1b71231e579fdeb388c5157636806b6b059a0R1485), so it's easier to tell what `2` and `3` are referring to.